tutorials/misc/migrating-to-0.4.0.md

```diff
@@ -0,0 +1,139 @@
+    instances: List[Instance] = []
+    # open some data_file
+    for line in data_file:
+       instance = ...
+       instances.append(instance)
+    return Dataset(instances)
+```
+
+In 0.4.0 the `read()` function is already implemented in the base `DatasetReader`
+class in a way that allows every dataset reader to generate instances either
+_lazily_ (that is, as needed, and anew each iteration) or
+_eagerly_ (that is, generated once and stored in a list)
+in a way that's transparent to the user.
+
+So in 0.4.0 your `DatasetReader` subclass needs to instead
+override the private `_read` function
+and `yield` instances one at time.
+
+This should only be like a 2-line change:
+
+```python
+# new name _read and return type: Iterable[Instance]
+def _read(self, file_path: str) -> Iterable[Instance]:
+    # open some data_file
+    for line in data_file:
+       instance = ...
+       # yield instances rather than collecting them in a list
+       yield instance
+```
+
+In addition, the base `DatasetReader` constructor now takes a `lazy: bool` parameter,
+which means that your subclass constructor should also take that parameter
+(unless you don't want to allow laziness, but why wouldn't you?)
+and explicitly pass it the superclass constructor:
+
+```python
+class MyDatasetReader(DatasetReader)
+    def __init__(self,
+                 # other arguments
+                 lazy: bool = False):
+        super().__init__(lazy)
+        # whatever other initialization you need
+```
+
+(For the reasoning behind this change, see the [Laziness tutorial](https://github.com/allenai/allennlp/blob/master/tutorials/getting_started/laziness.md).)
+
+# CHANGES YOU ARE MUCH LESS LIKELY TO RUN INTO
+
+If you only ever use our the command line tools (`python -m allennlp.run ...`) to train / evaluate models,
+you shouldn't have to change anything else. However, if you've written your own training loop,
+(or if you're just really interested in how AllenNLP works), there are a few more changes you should know about.
+
+## `Dataset`
+
+AllenNLP 0.4.0 gets rid of the `Dataset` class / abstraction.
+Previously, a `Dataset` was a wrapper for a concrete list of instances
+that contained logic for _indexing_ them with some vocabulary,
+and for converting them into `Tensor` s. And, as mentioned above,
+previously `DatasetReader.read()` returned such a dataset.
+
+In 0.4.0, `DatasetReader.read()` returns an `Iterable[Instance]`,
+which could be a list of instances or could produce them lazily.
+This means that the indexing and tensorization needs to happen elsewhere.
+
+## `Vocabulary`
+
+As `Dataset` no longer exists, we replaced `Vocabulary.from_dataset()`
+with `Vocabulary.from_instances()`, which accepts an `Iterable[Instance]`.
+In particulary, you'd most likely call this with the results of one or more calls
+to `DatasetReader.read()`.
+
+## `Batch`
+
+To handle tensorization,
+0.4.0 introduces the notion of a `Batch`,
+which is basically just a list of `Instance`s.
```

Comment:
Above you have `Tensor` s (with a space) and here you have it with no space.  Not sure which is right.